vowpalwabbit/core/src/experimental/reductions/ftrl.cc

```diff
@@ -0,0 +1,745 @@
+#include "vw/core/reductions/gd.h"
+#include "vw/core/setup_base.h"
+#include "vw/core/shared_data.h"
+#include "vw/core/simple_label.h"
+#include "vw/io/logger.h"
+#include "vw/serialization/type_constructors.h"
+
+#include <cfloat>
+#include <cmath>
+#include <string>
+
+using namespace VW::LEARNER;
+using namespace VW::config;
+using namespace VW::math;
+
+#define W_XT 0  // current parameter
+#define W_ZT 1  // in proximal is "accumulated z(t) = z(t-1) + g(t) + sigma*w(t)", in general is the dual weight vector
+#define W_G2 2  // accumulated gradient information
+#define W_MX 3  // maximum absolute value
+#define W_WE 4  // Wealth
+#define W_MG 5  // maximum gradient
+
+namespace
+{
+  /*
+class ftrl_update_data
+{
+public:
+  float update = 0.f;
+  float ftrl_alpha = 0.f;
+  float ftrl_beta = 0.f;
+  float l1_lambda = 0.f;
+  float l2_lambda = 0.f;
+  float predict = 0.f;
+  float normalized_squared_norm_x = 0.f;
+  float average_squared_norm_x = 0.f;
+};
+
+
+
+class ftrl
+{
+public:
+  VW::workspace* all = nullptr;  // features, finalize, l1, l2,
+  float ftrl_alpha = 0.f;
+  float ftrl_beta = 0.f;
+  ftrl_update_data data;
+  size_t no_win_counter = 0;
+  size_t early_stop_thres = 0;
+  uint32_t ftrl_size = 0;
+  std::vector<VW::reductions::details::gd_per_model_state> gd_per_model_states;
+};
+*/
+struct ftrl_update_data
+{
+  typesys::Prop<float> update;
+  typesys::Prop<float> ftrl_alpha;
+  typesys::Prop<float> ftrl_beta;
+  typesys::Prop<float> l1_lambda;
+  typesys::Prop<float> l2_lambda;
+  typesys::Prop<float> predict;
+  typesys::Prop<float> normalized_squared_norm_x;
+  typesys::Prop<float> average_squared_norm_x;
+};
+
+struct ftrl_data : model_data::versioned_data<1>
+{
+private:
+  using gd_model_state = VW::reductions::details::gd_per_model_state;
+  using persist_mode = model_data::persist_mode;
+  // TODO: The rest of the SGD Owl
+public:
+  class v1 : version<1>
+  {
+  private:
+    using gd_model_state = VW::reductions::details::gd_per_model_state;
+  public:
+    // ftrl mode options
+    model_data::container<bool> ftrl_enabled = persisted<bool, persist_mode::always>("ftrl").finish();
+    model_data::container<bool> pistol_enabled = persisted<bool, persist_mode::always>("pistol").finish();
+    model_data::container<bool> coin_enabled = persisted<bool, persist_mode::always>("coin").finish();
+    
+    model_data::container<ftrl_update_data> data = transient<ftrl_update_data>("ftrl_update_data").finish();
+
+    model_data::container<size_t> no_win_counter = transient<size_t>("no_win_counter").finish();
+    model_data::container<float> ftrl_alpha = transient<float>("ftrl_alpha").finish();
+    model_data::container<float> ftrl_beta = transient<float>("ftrl_beta").finish();
+    model_data::container<size_t> early_stop_thres = transient<size_t>("early_stop_thres").finish();
+    model_data::container<uint32_t> ftrl_size = persisted<uint32_t, persist_mode::save_resume>("ftrl_size").finish();
```

Comment:
This is _used_, but not actually _saved_ in save_load. Should it be persisted or transient?